app/server/lib/scim/v2/ScimUserUtils.ts

```diff
@@ -0,0 +1,46 @@
+import { normalizeEmail } from "app/common/emails";
+import { UserProfile } from "app/common/LoginSessionAPI";
+import { User } from "app/gen-server/entity/User.js";
+import SCIMMY from "scimmy";
+
+/**
+ * Converts a user from your database to a SCIMMY user
+ */
+export function toSCIMMYUser(user: User) {
+  if (!user.logins) {
+    throw new Error("User must have at least one login");
+  }
+  const locale = user.options?.locale ?? "en";
+  return new SCIMMY.Schemas.User({
+    id: String(user.id),
+    userName: user.loginEmail,
+    displayName: user.name,
+    name: {
+      formatted: user.name,
+    },
+    locale,
+    preferredLanguage: locale, // Assume preferredLanguage is the same as locale
+    photos: user.picture ? [{
+      value: user.picture,
+      type: "photo",
+      primary: true
+    }] : undefined,
+    emails: [{
+      value: user.logins[0].displayEmail,
+      primary: true,
+    }],
+  });
+}
+
+export function toUserProfile(scimUser: any, existingUser?: User): UserProfile {
+  const emailValue = scimUser.emails?.[0]?.value;
+  if (emailValue && normalizeEmail(emailValue) !== normalizeEmail(scimUser.userName)) {
+    throw new SCIMMY.Types.Error(400, 'invalidValue', 'Email and userName must be the same');
```

Comment:
I tested it against the [SCIM plugin for Keycloak](https://forge.libre.sh/libre.sh/keycloak-scim) (currently being developed). Of course you can have users with a userName and an email different.

A look at the RFCs have not revealed me that the userName is required to be taken into consideration by the service provider, despite being mandatory. I would probably just log a warning telling that both values are different and that the userName is ignored in such case.